src/server/semester/semester.service.ts

```diff
@@ -59,4 +72,160 @@ export class SemesterService {
 
       );
   }
+
+  /**
+   * Creates a new academic year of course instances, non class events, and
+   * faculty absences using the values of the previous semester.
+   * Retired course instances retain their "retired" status. Meetings and
+   * enrollment values are not retained in the newly created academic year.
+   */
+  public async addAcademicYear(newAcademicYear: number): Promise<void> {
+    // Note that academicYear in the semester table is actually calendar year.
+    const existingYears = await this.getYearList();
+    // If the academic year previous to the newAcademicYear does not exist,
+    // throw an exception.
+    if (!existingYears.includes((newAcademicYear - 1).toString())) {
+      throw new BadRequestException(
+        'Cannot create requested academic year until preceding academic year is created.'
+      );
+    }
+    // Only create the new academic year if it does not yet exist.
+    if (!existingYears.includes((newAcademicYear).toString())) {
+      let existingFallSemester: Semester;
+      let existingSpringSemester: Semester;
+      this.logService.info(`Creating academic year ${newAcademicYear}`);
```

Comment:
I probably should have given a little detail about the kind of logging we want to include. The different methods exposed by our `LogService` correspond to different streams of logs we can turn on or off at runtime -- i.e. locally we can log everything, but in production we'd log a smaller subset. Given that, it's worth being judicious in what gets logged to which stream. Here, I'd say we probably only want to log minimal beginning and ending message to the `info` stream, e.g.


```ts
  this.logService.info(`Creating academic year ${newAcademicYear}`);
  this.logService.info(`Successfully created academic year ${newAcademicYear}.`);
```

The other "process" messages could go into the `verbose` stream, which we might turn on in our QA environment. It's also usually a good idea to log out the number of items being created as you go (a la our ETL process). e.g.,

```ts
  this.logService.verbose('Creating the fall course instances.');
  this.logService.verbose(`Created ${fallInstanceCount} fall course instances.`);
  this.logService.verbose('Creating the fall non-class events.');
  this.logService.verbose(`Created ${fallEventCount} fall non-class events.`);
  this.logService.verbose('Creating the fall absences.');
  this.logService.verbose(`Created ${fallAbsenceCount} fall absences.`);
  this.logService.verbose('Creating the spring course instances.');
  this.logService.verbose(`Created ${springInstanceCount} spring course instances.`);
  this.logService.verbose('Creating the spring non-class events.');
  this.logService.verbose(`Created ${springEventCount} spring non-class events.`);
  this.logService.verbose('Creating the spring absences.');
  this.logService.verbose(`Created ${springAbsenceCount} spring absences.`);
```

Finally, we can add `debug` messages, which might dump out the full contents of the data being created. We would probably never use that in an AWS environment, but it could be useful when we're developing locally. So alongside the `verbose` logs above, you could add something like:

```ts
  this.logService.debug(fallSemester.courseInstances);
  this.logService.debug(fallSemester.absences);
  this.logService.debug(fallSemester.nonClassEvents);
  this.logService.debug(springSemester.courseInstances);
  this.logService.debug(springSemester.absences);
  this.logService.debug(springSemester.nonClassEvents);
```

One way to think about it: anywhere you would add a `console.log` while debugging, you can add a `logService.debug` or `logService.verbsose` call. Then when we run in production we can just ignore those streams all together.

Reply:
Yes, I do think that would be useful to keep in there. I'm imagining a situation where we're looking at the logs a year from now, when this isn't fresh, and it could be confusing to see the initial "creating new academic year" message without any additional resolution for the cases when it has already been create.

Reply:
Good point - better to have this note for us in the future.

Reply:
Logging was updated in [f907d0d](https://github.com/seas-computing/course-planner/pull/496/commits/f907d0ddece5f29dd65affca3bf1517853bd32b5).